File: chain/zul-privacy/src/instruction.rs
//! Shielded pool instruction wire format.
//!
//! The pool is an enshrined builtin program. Instructions are bincode of
//! `PoolInstruction`. The first account is always the pool state account;
//! value-moving instructions also reference the pool vault and the user's
//! token/lamport accounts (resolved by the node's privacy processor).

use serde::{Deserialize, Serialize};

use crate::field::FieldBytes;

/// Asset reference at the public boundary (shield / unshield).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum Asset {
    /// Native ZUL (lamports).
    Native,
    /// SPL token identified by mint (e.g. wrapped SOL on L2).
    Spl([u8; 32]),
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum PoolInstruction {
    /// Move public funds into the pool, creating one note commitment.
    /// `amount` of `asset` is debited from the depositor's public account.
    /// The node computes the commitment as
    /// `Poseidon3(asset_id, amount, secret)` so the committed value is bound
    /// to the public amount — the depositor only supplies the opaque
    /// `secret = Poseidon2(owner_pk, blinding)`.
    Shield {
        asset: Asset,
        amount: u64,
        secret: FieldBytes,
        /// Encrypted note payload for the recipient (ECDH + AEAD).
        encrypted_note: Vec<u8>,
    },
    /// Private 2-in/2-out transfer. Consumes two nullifiers, creates two
    /// commitments. No public value moves; an optional public fee is paid
    /// out of the pool to the fee collector.
    Transfer {
        proof: Vec<u8>,
        root: FieldBytes,
        nullifiers: [FieldBytes; 2],
        commitments: [FieldBytes; 2],
        fee: u64,
        encrypted_notes: [Vec<u8>; 2],
    },
    /// Spend a note to a public recipient. Consumes one nullifier, creates
    /// one change commitment, releases `amount` of `asset` publicly.
    Unshield {
        proof: Vec<u8>,
        root: FieldBytes,
        nullifier: FieldBytes,
        change_commitment: FieldBytes,
        asset: Asset,
        amount: u64,
        /// Bound into the proof's public inputs to stop front-running.
        recipient: [u8; 32],
        encrypted_change_note: Vec<u8>,
    },
}

impl PoolInstruction {
    pub fn try_from_slice(data: &[u8]) -> Result<Self, InstructionError> {
        bincode::deserialize(data).map_err(|_| InstructionError::Malformed)
    }

    pub fn to_bytes(&self) -> Vec<u8> {
        bincode::serialize(self).expect("instruction serialization is infallible")
    }
}

#[derive(Debug, thiserror::Error, PartialEq, Eq)]
pub enum InstructionError {
    #[error("malformed pool instruction")]
    Malformed,
}

impl Asset {
    pub fn is_native(&self) -> bool {
        matches!(self, Asset::Native)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn instruction_roundtrips() {
        let ix = PoolInstruction::Shield {
            asset: Asset::Native,
            amount: 1_000,
            secret: [3u8; 32],
            encrypted_note: vec![1, 2, 3],
        };
        let bytes = ix.to_bytes();
        assert_eq!(PoolInstruction::try_from_slice(&bytes).unwrap(), ix);

        let transfer = PoolInstruction::Transfer {
            proof: vec![9u8; 256],
            root: [1u8; 32],
            nullifiers: [[2u8; 32], [3u8; 32]],
            commitments: [[4u8; 32], [5u8; 32]],
            fee: 5_000,
            encrypted_notes: [vec![0u8; 80], vec![1u8; 80]],
        };
        let bytes = transfer.to_bytes();
        assert_eq!(PoolInstruction::try_from_slice(&bytes).unwrap(), transfer);
    }

    #[test]
    fn rejects_garbage() {
        assert_eq!(
            PoolInstruction::try_from_slice(&[0xff, 0xff, 0xff, 0xff]),
            Err(InstructionError::Malformed)
        );
    }
}

File: sdk/src/instruction.ts
// PoolInstruction encoding, byte-compatible with the Rust serde/bincode
// representation in chain/zul-privacy/src/instruction.rs.
//
// Enum layout (bincode v1): u32-le variant index, then fields in order.
//   Asset:  0 = Native | 1 = Spl([u8;32])
//   PoolInstruction: 0 = Shield | 1 = Transfer | 2 = Unshield

import { BincodeReader, BincodeWriter } from "./bincode.js";
import { frToBytes } from "./field.js";

export type Asset =
  | { kind: "native" }
  | { kind: "spl"; mint: Uint8Array };

export const Asset = {
  native(): Asset {
    return { kind: "native" };
  },
  spl(mint: Uint8Array): Asset {
    if (mint.length !== 32) throw new Error("mint must be 32 bytes");
    return { kind: "spl", mint };
  },
};

function writeAsset(w: BincodeWriter, asset: Asset): void {
  if (asset.kind === "native") {
    w.u32(0);
  } else {
    w.u32(1).fixedBytes(asset.mint, 32);
  }
}

function readAsset(r: BincodeReader): Asset {
  const tag = r.u32();
  if (tag === 0) return { kind: "native" };
  if (tag === 1) return { kind: "spl", mint: r.fixedBytes(32) };
  throw new Error(`invalid asset tag ${tag}`);
}

export interface ShieldArgs {
  asset: Asset;
  amount: bigint;
  // Opaque secret = Poseidon2(pk, blinding); the node derives the commitment
  // from the public (asset, amount) and this, binding the deposited amount.
  secret: Uint8Array; // 32-byte LE field
  encryptedNote: Uint8Array;
}

export interface TransferArgs {
  proof: Uint8Array; // 256-byte BE Groth16 blob
  root: Uint8Array;
  nullifiers: [Uint8Array, Uint8Array];
  commitments: [Uint8Array, Uint8Array];
  fee: bigint;
  encryptedNotes: [Uint8Array, Uint8Array];
}

export interface UnshieldArgs {
  proof: Uint8Array;
  root: Uint8Array;
  nullifier: Uint8Array;
  changeCommitment: Uint8Array;
  asset: Asset;
  amount: bigint;
  recipient: Uint8Array; // 32-byte pubkey
  encryptedChangeNote: Uint8Array;
}

export function encodeShield(args: ShieldArgs): Uint8Array {
  const w = new BincodeWriter().u32(0);
  writeAsset(w, args.asset);
  w.u64(args.amount);
  w.fixedBytes(args.secret, 32);
  w.vecU8(args.encryptedNote);
  return w.finish();
}

export function encodeTransfer(args: TransferArgs): Uint8Array {
  const w = new BincodeWriter().u32(1);
  w.vecU8(args.proof);
  w.fixedBytes(args.root, 32);
  w.fixedBytes(args.nullifiers[0], 32).fixedBytes(args.nullifiers[1], 32);
  w.fixedBytes(args.commitments[0], 32).fixedBytes(args.commitments[1], 32);
  w.u64(args.fee);
  w.vecU8(args.encryptedNotes[0]).vecU8(args.encryptedNotes[1]);
  return w.finish();
}

export function encodeUnshield(args: UnshieldArgs): Uint8Array {
  const w = new BincodeWriter().u32(2);
  w.vecU8(args.proof);
  w.fixedBytes(args.root, 32);
  w.fixedBytes(args.nullifier, 32);
  w.fixedBytes(args.changeCommitment, 32);
  writeAsset(w, args.asset);
  w.u64(args.amount);
  w.fixedBytes(args.recipient, 32);
  w.vecU8(args.encryptedChangeNote);
  return w.finish();
}

/// Decode helper (mainly for round-trip tests).
export function decodeInstructionTag(bytes: Uint8Array): number {
  return new BincodeReader(bytes).u32();
}

export { readAsset, writeAsset };

// Convenience: build a field-bytes commitment from a bigint.
export function commitmentBytes(value: bigint): Uint8Array {
  return frToBytes(value);
}
